compiler/expressions.go

```diff
@@ -1058,6 +1058,12 @@ func (fc *funcContext) translateBuiltin(name string, sig *types.Signature, args
 	case "Offsetof":
 		sel, _ := fc.selectionOf(astutil.RemoveParens(args[0]).(*ast.SelectorExpr))
 		return fc.formatExpr("%d", typesutil.OffsetOf(sizes32, sel))
+	case "SliceData":
+		// SliceData returns nil if the slice is nil, otherwise returns a pointer to the first index of the array, &s[0].
+		// If the slice is empty (cap == 0), it returns an "unspecified memory address" or in our case, a pointer to the empty array.
+		t := fc.typeOf(args[0]).Underlying().(*types.Slice)
+		elemPtrType := types.NewPointer(t.Elem())
+		return fc.formatExpr("(%1e === %2s.nil) ? %3s.nil : $indexPtr(%1e.$array, %1e.$offset, %3s)", args[0], fc.typeName(t), fc.typeName(elemPtrType))
```

Comment:
I think you need a temporary variable here for the `%1e` here, otherwise if it's a complex expression it would be evaluated multiple times, for example: `unsafe.SliceData(readLinesFromFile())` would insert the `readLinesFromFile()` call in every `%1e` location.

Reply:
Oh, great call, I didn't realize it did that for each instance of `%1e`. I'll fix that.

Reply:
Fixed it so that the `%e` formatting is done prior to the return and the formatting for the return uses `%1s` instead.

Reply:
Sorry, when I mentioned a temporary variable, I should've clarified I meant a JS variable. Unfortunately, with your fix the expression would still be evaluated multiple times, since the compiled expression would be inserted into the generated code several times.

However, on a second thought I think a cleaner option would be to define a `$sliceData` helper function in prelude and insert its call instead of the complex expression. This would both generate a more compact code, as well as avoid a multiple evaluation problem. You can see an example of that with the `close()` builtin a few lines above.

Reply:
Awesome, I'll give that a try.